AppCenterIdentity/AppCenterIdentity/MSIdentity.m

```diff
@@ -310,4 +305,60 @@ - (void)removeAuthToken {
   }
 }
 
+- (void)acquireTokenSilentlyWithMSALAccount:(MSALAccount *)account {
+  __weak typeof(self) weakSelf = self;
+  [self.clientApplication
+      acquireTokenSilentForScopes:@[ (NSString * _Nonnull) self.identityConfig.identityScope ]
+                          account:account
+                  completionBlock:^(MSALResult *result, NSError *e) {
+                    typeof(self) strongSelf = weakSelf;
+                    if (e) {
+                      MSLogWarning([MSIdentity logTag],
+                                   @"Silent acquisition of token failed with error: %@. Triggering interactive acquisition", e);
+                      [strongSelf acquireTokenInteractively];
+                    } else {
+                      [MSAuthTokenContext sharedInstance].authToken = result.idToken;
+                      [strongSelf saveAuthToken:result.idToken];
+                      [strongSelf saveAccountId:(NSString * _Nonnull) result.account.homeAccountId];
+                    }
+                  }];
+}
+
+- (void)acquireTokenInteractively {
+  __weak typeof(self) weakSelf = self;
+  [self.clientApplication acquireTokenForScopes:@[ (NSString * _Nonnull) self.identityConfig.identityScope ]
+                                completionBlock:^(MSALResult *result, NSError *e) {
+                                  if (e) {
+                                    MSLogError([MSIdentity logTag], @"User sign-in failed. Error: %@", e);
+                                  } else {
+                                    typeof(self) strongSelf = weakSelf;
+                                    [MSAuthTokenContext sharedInstance].authToken = result.idToken;
+                                    [strongSelf saveAuthToken:result.idToken];
+                                    [strongSelf saveAccountId:(NSString * _Nonnull) result.account.homeAccountId];
```

Comment:
Are you sure this conversion is ok? `homeAccountId` has a type with three different properties.

Reply:
I guess you should use `homeAccountId.identifier` here, it is declared as `nonnull` so you also won't have to do this `nullable` to `nonnull` conversion which looks super-unsafe to me.

Reply:
@annakocheshkova good catch, will fix this.

Reply:
@dhei I partly fixed this in my branch too, (to be able to test), but I guess it will not produce serious merge conflicts.